base/compiler/typeinfer.jl

```diff
@@ -821,6 +817,9 @@ function resolve_call_cycle!(interp::AbstractInterpreter, mi::MethodInstance, pa
                 poison_callstack(parent, frame)
                 return true
             end
+            # Also taint the termination effect, because we can no longer guarantee the absence
+            # of recursion.
+            merge_effects!(interp, parent, Effects(EFFECTS_TOTAL; terminates=false))
```

Comment:
I don't think this is needed, since it should get tainted in `abstract_call_method` when the call fails to return without hitting a cycle.